### DISCUSSION TITLE
Redirect external app std::cout to _console?

### DISCUSSION THREAD

**Discussion by aprilnovak** [idaholab/moose#16647] [Q&A General] [answered]
Hi all,

I'm running an external app (nekRS) through the `ExternalProblem` interface, and it would be nice to redirect nekRS's `std::cout` through MOOSE's `_console` to make the output prettier. 

For instance, nekRS uses `std::cout`, which when run as a sub-app, looks like this on my console. Lines 4 and 5 come from nekRS.

**nek0:** Time Step 17, time = 0.029, dt = 0.002
**nek0:** Sending heat flux to nekRS boundary 1...
**nek0:** Normalizing total nekRS flux of 389.376 to the conserved MOOSE value of 386.989...
step= 17  t= 2.90000000e-02  dt=2.0e-03  C= 0.05  UVW: 6  P: 201  S: 157  eTime= 6.60e-01, 1.08991e+01 s
copying solution to nek
**nek0:** Extracting nekRS temperature from boundary 1...
**nek0:** Interpolated temperature min/max values on interface: 628.15, 628.952
**nek0:**  Solve Converged!

Does MOOSE have a capability like this? If this is obvious, my apologies - nothing jumped out at me on `_console`!

Thanks!

-April

**Comment by aeslaughter** [comment, marked as answer]
I don't know how that could be done. @permcody do you know if this would be possible?

**Reply by permcody** [reply]
To the best of my knowledge a simple redirect definitely isn't possible. The `_console` is a C++ object that manages it's own stream. It may be possible to share the underlying stream but even that would take a bit of gymnastics. 

I think a better idea would be to add a new API (extern C linkage) that developers of other applications could call that would invoke the MOOSE console object. This would be fairly trivial to implement and would automatically give you the coloring, indentation, and input file based control we are accustomed to. The big disadvantage is that this would be very disruptive for applications that have output spread all over through their application as it would require them to go through their application and update everything. Worse yet, you'd have to have make this a macro to handle the cases when you are wrapped in MOOSE or not. 

Again there are ways of sharing file system streams at the OS level, but I'm not sure of a way we could put all of the MOOSE functionality into one of those and share it. It may be doable, but we'd need to build a demo project to try a few things out.

**Reply by aprilnovak** [reply]
Thanks for explaining everything! Unfortunately, nekRS has `cout` scattered over the entire project, so I agree that this would be pretty tricky to implement. And the nekRS devs are pretty against having any trace of "MOOSE" things in their project from a modularization perspective, understandably.

This was just something I thought would be nice for formatting, so this is very low priority anyways. Just wanted to check that that feature didn't exist yet!

Thanks!
